### ISSUE TITLE
Compilation error when raising CellVolume to a non-integer power

### ISSUE THREAD

**Issue by ctjacobs** [firedrakeproject/firedrake#490]
This might be related to #488, but I have updated FFC, Firedrake, PyOP2, etc, since then. The following minimal example fails:

```
from firedrake import *

mesh = UnitSquareMesh(10, 10)
fs = FunctionSpace(mesh, "CG", 1)
test = TestFunction(fs)
trial = TrialFunction(fs)

volume = CellVolume(mesh)**0.5

F = inner(test, trial)*dx - inner(test, volume)*dx
solution = Function(fs)
solve(lhs(F) == rhs(F), solution)
```

with:

```
pyop2:INFO Solving linear variational problem...
pyop2:INFO   Compiling wrapper...
Traceback (most recent call last):
  File "test.py", line 12, in <module>
    solve(lhs(F) == rhs(F), solution)
  File "/data/firedrake/firedrake/solving.py", line 119, in solve
    _solve_varproblem(*args, **kwargs)
  File "/data/firedrake/firedrake/solving.py", line 144, in _solve_varproblem
    solver.solve()
  File "<string>", line 2, in solve
  File "/usr/local/lib/python2.7/dist-packages/PyOP2-0.11.0_177_ga14e2e7_dirty-py2.7-linux-x86_64.egg/pyop2/profiling.py", line 197, in wrapper
    return f(*args, **kwargs)
  File "/data/firedrake/firedrake/variational_solver.py", line 153, in solve
    self.snes.solve(None, v)
  File "PETSc/SNES.pyx", line 516, in petsc4py.PETSc.SNES.solve (src/petsc4py.PETSc.c:161937)
  File "PETSc/petscsnes.pxi", line 251, in petsc4py.PETSc.SNES_Function (src/petsc4py.PETSc.c:30784)
  File "/data/firedrake/firedrake/solving_utils.py", line 181, in form_function
    with ctx._Fs[lvl].dat.vec_ro as v:
  File "/usr/lib/python2.7/contextlib.py", line 17, in __enter__
    return self.gen.next()
  File "/usr/local/lib/python2.7/dist-packages/PyOP2-0.11.0_177_ga14e2e7_dirty-py2.7-linux-x86_64.egg/pyop2/petsc_base.py", line 140, in vec_context
    self._force_evaluation()
  File "/usr/local/lib/python2.7/dist-packages/PyOP2-0.11.0_177_ga14e2e7_dirty-py2.7-linux-x86_64.egg/pyop2/base.py", line 1558, in _force_evaluation
    _trace.evaluate(reads, writes)
  File "/usr/local/lib/python2.7/dist-packages/PyOP2-0.11.0_177_ga14e2e7_dirty-py2.7-linux-x86_64.egg/pyop2/base.py", line 154, in evaluate
    comp._run()
  File "/usr/local/lib/python2.7/dist-packages/PyOP2-0.11.0_177_ga14e2e7_dirty-py2.7-linux-x86_64.egg/pyop2/base.py", line 3830, in _run
    return self.compute()
  File "<string>", line 2, in compute
  File "/usr/local/lib/python2.7/dist-packages/PyOP2-0.11.0_177_ga14e2e7_dirty-py2.7-linux-x86_64.egg/pyop2/profiling.py", line 197, in wrapper
    return f(*args, **kwargs)
  File "/usr/local/lib/python2.7/dist-packages/PyOP2-0.11.0_177_ga14e2e7_dirty-py2.7-linux-x86_64.egg/pyop2/base.py", line 3839, in compute
    self._compute(self.it_space.iterset.core_part)
  File "/usr/local/lib/python2.7/dist-packages/PyOP2-0.11.0_177_ga14e2e7_dirty-py2.7-linux-x86_64.egg/pyop2/sequential.py", line 154, in _compute
    fun(*self._jit_args, argtypes=self._argtypes, restype=None)
  File "/usr/local/lib/python2.7/dist-packages/PyOP2-0.11.0_177_ga14e2e7_dirty-py2.7-linux-x86_64.egg/pyop2/host.py", line 634, in __call__
    return self.compile(argtypes, restype)(*args)
  File "/usr/local/lib/python2.7/dist-packages/PyOP2-0.11.0_177_ga14e2e7_dirty-py2.7-linux-x86_64.egg/pyop2/host.py", line 737, in compile
    compiler=compiler.get('name'))
  File "/usr/local/lib/python2.7/dist-packages/PyOP2-0.11.0_177_ga14e2e7_dirty-py2.7-linux-x86_64.egg/pyop2/compilation.py", line 255, in load
    dll = compiler.get_so(src, extension)
  File "/usr/local/lib/python2.7/dist-packages/PyOP2-0.11.0_177_ga14e2e7_dirty-py2.7-linux-x86_64.egg/pyop2/compilation.py", line 131, in get_so
    Compile errors in %s""" % (e.cmd, e.returncode, logfile, errfile))
pyop2.exceptions.CompilationError: Command "['mpicc', '-std=c99', '-fPIC', '-Wall', '-g', '-O3', '-fno-tree-vectorize', '-I/usr/local/lib/python2.7/dist-packages/petsc/include', '-I/data/firedrake/firedrake', '-I/usr/local/lib/python2.7/dist-packages/PyOP2-0.11.0_177_ga14e2e7_dirty-py2.7-linux-x86_64.egg/pyop2', '-msse', '-o', '/tmp/pyop2-cache-uid1000/6fde57abd2bad82c98643f476f11ac57.so.tmp', '/tmp/pyop2-cache-uid1000/6fde57abd2bad82c98643f476f11ac57.c', '-shared', '-L/usr/local/lib/python2.7/dist-packages/petsc/lib', '-Wl,-rpath,/usr/local/lib/python2.7/dist-packages/petsc/lib', '-lpetsc', '-lm']" return error status 1.
Unable to compile code
Compile log in /tmp/pyop2-cache-uid1000/6fde57abd2bad82c98643f476f11ac57.log
Compile errors in /tmp/pyop2-cache-uid1000/6fde57abd2bad82c98643f476f11ac57.err
```

* [6fde57abd2bad82c98643f476f11ac57.log](https://gist.github.com/ctjacobs/9c16419d70edfb79a6e9)
* [6fde57abd2bad82c98643f476f11ac57.err](https://gist.github.com/ctjacobs/d5f3742b6e40d6df1707)

**Comment by dorugeber** [comment]
Paste the c file?

Does it work if you just have `volume = CellVolume(mesh)`?  What about `volume = CellVolume(mesh)**1.0`?

**Comment by ctjacobs** [comment]
[6fde57abd2bad82c98643f476f11ac57.c](https://gist.github.com/ctjacobs/504ebc63ac54c4576928)

Yes and yes.

**Comment by dorugeber** [comment]
Does `volume = CellVolume(mesh)**2.0` work?  If so, can you paste the generated code?

**Comment by ctjacobs** [comment]
`volume = CellVolume(mesh)**2.0` does not work, but `volume = CellVolume(mesh)**2` does. Here's the .c file for `volume = CellVolume(mesh)**2.0`: https://gist.github.com/ctjacobs/b4fca07b6800d0f320e2 and for `volume = CellVolume(mesh)**2`: https://gist.github.com/ctjacobs/2b1364ea958f16a5b80f

**Comment by dorugeber** [comment]
Yep, as I guessed, fecking `remove_unused`.

As a stop-gap measure, can I suggest that you change line 874 of `ffc/cpp.py` to `    return code`, thereby bypassing that function?

**Comment by ctjacobs** [comment]
Thanks. That works around the issue.

**Comment by wence-** [comment]
DId we have an actual solution to this problem?

**Comment by ctjacobs** [comment]
I don't think so. Just checked with up-to-date FFC, PETSc, etc and I'm still getting the error without the work-around.

**Comment by wence-** [comment]
Still borked, still annoying.

**Comment by wence-** [comment]
Fixed in #691
